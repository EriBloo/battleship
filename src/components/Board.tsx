import React, { ReactElement, useState, useEffect } from 'react';
import Battleship from '../scripts/Battleship';
import Game from '../scripts/Game';
import '../styles/Board.scss';

function Board(props: {
  player: 0 | 1;
  game: Game;
  state: { [state: string]: [number, number][] };
  loop: (loc: [number, number]) => void;
}): ReactElement {
  const [currentTurn, setCurrentTurn] = useState(props.game.getTurn);
  const [active, setActive] = useState('');

  function updateTiles(): void {
    if (props.player === 0) {
      props.state.shipNotHit.map((ship) => {
        const element = document.querySelector(
          `.board-tile[data-x="${ship[0]}"][data-y="${ship[1]}"][data-player="${props.player}"]`,
        );
        element?.classList.add('ship-not-hit');
      });
    }
    props.state.shipHit.map((ship) => {
      const element = document.querySelector(
        `.board-tile[data-x="${ship[0]}"][data-y="${ship[1]}"][data-player="${props.player}"]`,
      );
      if (
        (props.game.getPlayer(props.player).getBoard.getTiles[ship[0]][
          ship[1]
        ] as Battleship).isSunk()
      ) {
        element?.classList.add('ship-sunk');
      } else {
        element?.classList.add('ship-hit');
      }
    });
    props.state.missed.map((ship) => {
      const element = document.querySelector(
        `.board-tile[data-x="${ship[0]}"][data-y="${ship[1]}"][data-player="${props.player}"]`,
      );
      element?.classList.add('missed');
    });
  }

  function loop(e: React.MouseEvent<HTMLDivElement>): void {
    if (props.player === 1 - currentTurn) {
      const el = e.target as HTMLDivElement;
      props.loop([parseInt(el.getAttribute('data-x') as string, 10), parseInt(el.getAttribute('data-y') as string, 10)]);
    }
  }

  useEffect(() => {
    if (currentTurn === 1 - props.player) {
      setActive('active');
    } else {
      setActive('');
    }
  }, [currentTurn]);

  useEffect(() => {
    updateTiles();
  }, [props.state]);

  return (
    <div className="board-wrapper">
      {props.game.getPlayer(props.player).getBoard.getTiles.map((row, i) => {
        return row.map((_, j) => {
          return (
            <div
              key={`(${i}, ${j})`}
              data-x={`${i}`}
              data-y={`${j}`}
              data-player={props.player}
              className={`board-tile ${active}`}
              onClick={loop}
            />
          );
        });
      })}
    </div>
  );
}

export default Board;
