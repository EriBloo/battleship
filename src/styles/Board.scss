.board-wrapper {
  width: 20rem;
  height: 20rem;
  display: grid;
  grid-template-columns: repeat(10, 1fr);
  grid-template-rows: repeat(10, 1fr);
  background-color: #00adb5;
  border: 1px solid #00adb5;
  grid-gap: 1px;
  opacity: .6;

  .board-tile {
    background-color: #f4f4f4;
  }

  .active:hover {
    background-color: rgb(166, 214, 166);
    border: 2px solid rgb(92, 158, 92);
    cursor: pointer;
  }

  .ship-not-hit {
    background-color: rgb(109, 109, 168);
  }

  .ship-hit {
    background-color: rgb(194, 135, 135);
  }

  .ship-sunk {
    background-color: rgb(116, 116, 116);
  }

  .missed {
    background-color: rgb(236, 236, 162);
  }
}