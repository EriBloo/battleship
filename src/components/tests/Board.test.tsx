import React from 'react';
import renderer from 'react-test-renderer';
import Board from '../Board';
import Game from '../../scripts/Game';

test('component renders', () => {
  const game = new Game([2, 2]);
  const tree = renderer
    .create(
      <Board
        game={game}
        player={0}
        state={game.getCurrentPlayer.getBoard.getBoardStates}
        loop={() => {
          return;
        }}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
