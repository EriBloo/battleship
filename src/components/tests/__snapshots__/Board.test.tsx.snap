// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component renders 1`] = `
<div
  className="board-wrapper"
>
  <div
    className="board-tile "
    data-player={0}
    data-x="0"
    data-y="0"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="0"
    data-y="1"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="0"
    data-y="2"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="0"
    data-y="3"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="0"
    data-y="4"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="0"
    data-y="5"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="0"
    data-y="6"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="0"
    data-y="7"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="0"
    data-y="8"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="0"
    data-y="9"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="1"
    data-y="0"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="1"
    data-y="1"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="1"
    data-y="2"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="1"
    data-y="3"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="1"
    data-y="4"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="1"
    data-y="5"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="1"
    data-y="6"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="1"
    data-y="7"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="1"
    data-y="8"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="1"
    data-y="9"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="2"
    data-y="0"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="2"
    data-y="1"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="2"
    data-y="2"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="2"
    data-y="3"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="2"
    data-y="4"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="2"
    data-y="5"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="2"
    data-y="6"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="2"
    data-y="7"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="2"
    data-y="8"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="2"
    data-y="9"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="3"
    data-y="0"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="3"
    data-y="1"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="3"
    data-y="2"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="3"
    data-y="3"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="3"
    data-y="4"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="3"
    data-y="5"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="3"
    data-y="6"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="3"
    data-y="7"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="3"
    data-y="8"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="3"
    data-y="9"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="4"
    data-y="0"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="4"
    data-y="1"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="4"
    data-y="2"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="4"
    data-y="3"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="4"
    data-y="4"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="4"
    data-y="5"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="4"
    data-y="6"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="4"
    data-y="7"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="4"
    data-y="8"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="4"
    data-y="9"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="5"
    data-y="0"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="5"
    data-y="1"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="5"
    data-y="2"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="5"
    data-y="3"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="5"
    data-y="4"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="5"
    data-y="5"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="5"
    data-y="6"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="5"
    data-y="7"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="5"
    data-y="8"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="5"
    data-y="9"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="6"
    data-y="0"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="6"
    data-y="1"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="6"
    data-y="2"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="6"
    data-y="3"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="6"
    data-y="4"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="6"
    data-y="5"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="6"
    data-y="6"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="6"
    data-y="7"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="6"
    data-y="8"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="6"
    data-y="9"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="7"
    data-y="0"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="7"
    data-y="1"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="7"
    data-y="2"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="7"
    data-y="3"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="7"
    data-y="4"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="7"
    data-y="5"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="7"
    data-y="6"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="7"
    data-y="7"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="7"
    data-y="8"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="7"
    data-y="9"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="8"
    data-y="0"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="8"
    data-y="1"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="8"
    data-y="2"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="8"
    data-y="3"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="8"
    data-y="4"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="8"
    data-y="5"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="8"
    data-y="6"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="8"
    data-y="7"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="8"
    data-y="8"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="8"
    data-y="9"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="9"
    data-y="0"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="9"
    data-y="1"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="9"
    data-y="2"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="9"
    data-y="3"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="9"
    data-y="4"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="9"
    data-y="5"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="9"
    data-y="6"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="9"
    data-y="7"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="9"
    data-y="8"
    onClick={[Function]}
  />
  <div
    className="board-tile "
    data-player={0}
    data-x="9"
    data-y="9"
    onClick={[Function]}
  />
</div>
`;
